README
    Oracle Information Site 2.04 - Readme (Last updated on 01-Feb-2015)
-----------------------------------------------------------------------------
This program is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

This document provides installation instructions and a short overview of the
integrated features.

1. INTRODUCTION
~~~~~~~~~~~~~~~
OIS 2.xx is the successor of my OIS 1.x application. The old OIS was assembled
in the early days of 2001 with table layouts, hard-coded styles and limited
functionality to extend OIS with additional plugins.

So in 2009 I've decided to rewrite the whole application from scratch using
Web 2.0 functionality (i hate this term) and a flexible plugin structure
which allows other developers to enhance OIS2 with new functions.
In fact, all functionality provided by OIS 1.x has been integrated in OIS 2.xx
as plugins.

The major change to OIS 1.xx is the way you are connecting to Oracle. The
previous version required a hard-coded setup for username, password and TNS
name to connect to the databases. This is a major security problem as the
passwords of the users have to be entered in plain text, which makes it easy
for others to get the credentials.

OIS 2.xx no longer stores any usernames and passwords. Instead you are entering
these informations on the login page and OIS 2.xx authenticates your
crendentials against the choosen database. This way the passwords are no
longer stored in the configuration and your database privileges decides what
you can view from the various V$ and DBA_ views Oracle offers. Even a connect
with the well-known user "SCOTT" is now possible, however during the limited
privileges of that user you won't be able to see much database
informations at all.


2. REQUIREMENTS
~~~~~~~~~~~~~~~
OIS 2.xx requires the following software components to work:

- At least one Oracle instance, supported releases are 8i up to 11gR2

- A decent browser capable of Javascript and CSS 2.1. I've tested with
  Firefox 35.x, IE10, Opera 12, Safari 5.x, Chrome and iCab.

- PHP 5.x / PHP 7.x / PHP 8.0.x

- PHP components: OCI8, SESSION, JSON

- Optional the JPGraph classes (see http://www.aditus.nu/jpgraph/index.php)
  NOTE: Currently only the 3.x series is correctly supported!

- To connect as SYSDBA/SYSOPER you have to set the PHP init variable
  "oci8.privileged_connect" to "on" (only for PHP 5.2+)

- jQuery + jQuery UI is already contained in the release archive


3. INSTALLATION
~~~~~~~~~~~~~~~
Installation of OIS2 is rather simple. Extract the archive you've downloaded
to a location where your Webserver document root is located. The archive
extracts to an own directory called OIS2.

If you just want to have a first look point your browser to the OIS2 directory
of your webserver, no further configuration is required to use OIS. Just login
to your selected database with the user you want. The TNS names are
automatically read and shown in the login window.

OIS allows to fine-tune the installation by providing a configuration file.
The installation archive supplies the file "inc/config.inc.php.sample", just
rename it to "inc/config.inc.php" and adapt the settings to match your
environment.

Here's a brief description of the available config options:

-------------------------------------------------------------------------------
OIS_INSTALL_PATH (STRING)
-------------------------------------------------------------------------------
This define specifies the real filesystem path where OIS2 is installed.
Only set this if autodetection does not work (and let me know WHY it
doesn't work for you!).


-------------------------------------------------------------------------------
OIS_INSTALL_URL (STRING)
-------------------------------------------------------------------------------
This define specifies the URL to reach OIS2 via your Web-Browser. As OIS2
auto-detects where it is running you should define this only if OIS2 was not
able to auto-detect the correct value. Again please give me a short mail in
case the auto-detection does not work.


-------------------------------------------------------------------------------
UI_THEME (STRING)
-------------------------------------------------------------------------------
Enter here the name of a jQueryUI theme, default is "smoothness".
All Themes have to be located as separate directories under the "css"
directory of OIS.
To make it easier for users to upgrade or exchange a given theme, simply
download a new one from the jQueryUI page and extract the contents in an own
directory below the "css" directory, then change this option here with the
name of your new theme.


-------------------------------------------------------------------------------
$OIS_DATABASES (ARRAY)
-------------------------------------------------------------------------------
Enter here all TNS names of the Oracle instances you want to use together
with OIS2. You can keep this array empty, in this case OIS reads the
tnsnames.ora file from $ORACLE_HOME/network/admin or from the directory
given by environment variable $TNS_ADMIN.


-------------------------------------------------------------------------------
JPGRAPH_PATH (STRING)
-------------------------------------------------------------------------------
OIS2 supports dynamic creation of graphs with the help of the superb JPGraph
package, which you can download under http://jpgraph.net/.
If you want to use these classes, specify here the full path to your
extracted JPGraph package. The path must point to the directory where the
base class "jpgraph.php" is located.


4. USAGE
~~~~~~~~
Point your browser to the URL where OIS2 is reachable on your server. You
will see a login mask with a selection of all TNS names you've configured.

If PHP 5.2 or newer is used and you have also set the PHP ini option
"oci8.privileged_connect" to "on", you can choose the connection
type, which is one of "NORMAL", "SYSOPER" or "SYSDBA.

Simply enter your Oracle Username, the Oracle Password and choose the TNS
name to connect, then click on Login. OIS2 then tries to authenticate your
given credentials against the Oracle instance, and if all is okay, you will
be redirected to the main screen, where some database informations are shown
together with the list of installed plugins.


5. PLUGINS
~~~~~~~~~~
OIS2 implements all functionality via plugins and some plugins are already
included in the release archive. The following plugins are provided in 2.03:

--------------------------------------------------------------------------
* Home
--------------------------------------------------------------------------
Home displays some database informations and a list of installed plugins.


--------------------------------------------------------------------------
* DB Info
--------------------------------------------------------------------------
DB Info displays global database informations:

 - General Informations
 - Initialisation parameter
 - SGA/PGA usage (depends on used Oracle Release)
 - Instance statistics
 - Control file overview
 - Used features
 - Registry


--------------------------------------------------------------------------
* DB Jobs
--------------------------------------------------------------------------
Overview of all Oracle Scheduler jobs from both DBMS_JOB and DBMS_SCHEDULER.
Note that DBMS_SCHEDULER is available only for Oracle 10g or newer, while
DBMS_JOB is available since Oracle 8i and still exists in 11gR2.
The OIS2 code should auto-detect the used Oracle version and display always
data which is valid for the given Oracle version.
Since 2.01 also the views DBMS_SCHEDULER_JOB_RUN_DETAILS and
DBMS_SCHEDULER_PROGRAMS are displayed.


--------------------------------------------------------------------------
* Flashback
--------------------------------------------------------------------------
Overview of Flashback and Recycle Bin functionality. Flashback and the
recycle bin are new features introduced by Oracle 10g and newer, so for
older releases (8i/9i) you will get an error message if you start this plugin.


--------------------------------------------------------------------------
* Processes
--------------------------------------------------------------------------
This plugin displays a complete overview of all currently connected sessions.
Also it provides two views to show the "Top 20 CPU" and "Top 20 Wait"
processes. "Top 20 CPU" shows the Top 20 processes consuming most of the
available CPU time while "Top 20 Wait" lists the top 20 processes which have
the highest times "waiting" on something like latches etc.
Since V2.01 a new tab is added called "Long Ops". This tab will show long
running operations from V$SESSION_LONGOPS with TIME_REMAINING > 0. This
view is refreshed on a 3 second interval.
User sessions can be killed (if you have the necessary permissions) by
clicking on the red cross right to the process entry.


--------------------------------------------------------------------------
* Redo Logs
--------------------------------------------------------------------------
Displays a list of redo logs used by the current Oracle instance together
with some statistics about the usage of the redo logs.


--------------------------------------------------------------------------
* Rollback Seg.
--------------------------------------------------------------------------
Lists rollback segments from the current Oraclce instance. In Oracle 9i
and newer releases the UNDO tablespace informations are shown, too.


--------------------------------------------------------------------------
* Schema Viewer
--------------------------------------------------------------------------
Allows to view a list of objects available for a given schema. You can
also view the DDL of these objects if running Oracle 9i or newer. The
DDL is shown via the help of the DBMS_METADATA package which Oracle
introduced with Oracle 9i. Note that the DBMS_METADATA output is not
necessarly the same call you've made to actually create the object, Oracle
will transform the DDL call and add various options, even if they are
all at their default values.


--------------------------------------------------------------------------
* SQL Monitor
--------------------------------------------------------------------------
Displays all currently executing queries on your database instance in a
two second refresh delay. Useful to track workload of your database.


--------------------------------------------------------------------------
* Tablespaces
--------------------------------------------------------------------------
Displays an overview of all tablespaces defined in the current instance.
Also a "disc activity" table is shown which lists the various datafiles
together with their usage statistics, sorted by disc activity.
If you click on a name of a tablespace you will see the list of datafiles
the given tablespace defines.
Clicking on the Mag displays a list of all objects currently stored in
the tablespace.


--------------------------------------------------------------------------
* Top 20 Queries
--------------------------------------------------------------------------
Displays the Top 20 queries sorted by disc reads per execution. Also the
SQL statements with the most invalidations are shown here. You can click
on these SQL statements to view the details about the query performance.


--------------------------------------------------------------------------
* Users
--------------------------------------------------------------------------
Lists all registered users for the current instance. Click on an username
to view more details like assigned Roles etc.
In addition to the userlist you can view here also all defined roles and
profiles for the current instance.


--------------------------------------------------------------------------
* Logout
--------------------------------------------------------------------------
Logs you off from OIS2 and your database.



6. CONTACT
~~~~~~~~~~
If you have any suggestions, bug-reports or improvements for OIS2, feel free
to contact me on one of the methods listed below:

E-MAIL: php@saschapfalz.de
   WWW: http://www.saschapfalz.de/contact.php

Normally I answer to emails in less than 8 hours, if this is not the case I'm
either very busy or your mail was killed by a spam filter, in this case resend
your email please or use the contact form on my homepage.
-------------------------------------------------------------------------------


CHANGELOG
APPLICATION=Oracle Information Site 2

V2.04:

26-Sep-2021 - Public Release 2.04

            - Made code compatible with PHP 8

            - Updated db_oci8.class to 1.1.0

30-Jun-2018 - Fixed problems when fresh-install of OIS2 is used without
              configuration file

V2.03:

01-Feb-2015 - Public Release as 2.03

            - Updated README to reflect latest changes.

            - Updated sgl_functions.class to 0.23

31-Jan-2015 - Changed HTML output to use some HTML 5 features.

            - Updated jQueryUI to 1.11.2

            - Updated jQuery to 1.11.2

            - Made config.inc.php file obsolete, now all important defines are
              automatically set in nocache.inc.php. Only if logging should be
              used and/or jpgraph should be utilized the config.inc.php file
              must be provided.

V2.02:

21-Jul-2014 - Added new plugin "SQLMonitor"

            - Added global logging function WriteLog() which is configured
              via config.inc.php by specifing a filename for the logfile under
              the define "OIS_LOGFILE".

20-Jul-2014 - Updated plugin "Flashback" with new features.

18-Jul-2014 - Moved navigation JS code to separate file

17-Jul-2014 - Added additional search for tnsnames.ora from environment
              variable "TNS_ADMIN".

            - Updated sgl_functions.class to 0.21

            - Changed UI_THEME define, now this define holds only the name
              of the subfolder under "css" directory where you have put the
              theming of jqueryUI. Defaults to "smoothness"

            - Updated jQueryUI to 1.11.0

            - Updated jQuery to 1.11.1

29-Oct-2013 - Added tnsparser.class to allow automatic recognition of all
              TNS files found on the local database server.
              NOTE: Searches only on standard path, wich is :
              $ORACLE_HOME/network/admin/tnsnames.ora

30-May-2013 - Updated dbjobs plugin so that SCHEDULER jobs can be enabled or
              disabled by clicking on the state symbol.

            - Updated OCI8 class to 1.04

            - Updated sgl_functions class to 0.19

            - Updated jQuery to 1.9.1 and jQueryUI to 1.10.3, added also the
              jQuery-migrate-1.2.1 codebase for backward compability.

22-May-2013 - Updated "Show Processes" plugin to 2.02

V2.01:

02-Oct-2011 - Public release as 2.01

08-Sep-2011 - Updated jQuery to 1.6.3 and jQueryUI to 1.8.16

31-Aug-2011 - Updated DBInfo plugin.

19-Jul-2011 - Added new Tab "Feature usage" to DBInfo plugin, this plugin
              displays the view DBA_FEATURE_USAGE_STATISTICS (Only for
              Oracle 10g or higher)

            - Updated sgl_functions class to 0.16

06-Jul-2011 - Updated sgl_functions class to 0.15

30-Jun-2011 - Updated OCI8 class to 1.01

12-May-2011 - Updated jQueryUI to 1.8.11

            - Updated plugin DBInfo and added RESULT_CACHE informations when
              connected to an Oracle DB 11g or higher.

V2.00:

24-Dec-2010 - Added check for JSON* functionality in PHP on main index page.
              If json_encode() is not defined, the index page aborts with
              an error.

20-Dec-2010 - Updated jQueryUI and smoothness theme to 1.8.7 .

            - Changed global function InformUser() to use the jQueryUI theme
              to render the box, this hopefully works flawelessly with different
              themes.

            - Rewritten code to fully utilize the themes from jQueryUI.

27-Nov-2010 - Added possibility to auto-refresh all pages/tabs. Just choose the
              refresh rate below the navigation, then the refresh will be taken
              on either the currently selected page or tab, depends on what you
              are currently viewing. Requested by Marton Szabo.

07-Nov-2010 - Added CHANGELOG file to all plugins, this makes changelog tracking
              a lot easier.

            - Updated jQuery to 1.4.4, jQueryUI to 1.8.6 + new theme smoothness.

            - Changed OIS2EXT method "PrintExtHeader()" to always include
              the jQueryUI code + the CSS files. You can install additional
              themes for the jQueryUI parts by copying the theme files to
              css/[themename] and change the define UI_THEME to [themename]
              in the file inc/defines.inc.php.

15-Aug-2010 - Added list of all roles and profiles defined in the current
              database to the "showusers" plugin.

            - Added views "DBA_SYS_PRIVS","DBA_TAB_PRIVS","DBA_TS_QUOTAS"
              to user detail window.

            - Added new method "RenderQuery()" to OIS2EXT class, this
              method allows to easily dump out the result of a query
              in a configurable HTML table output. See the plugin
              "showusers" for details how to use this method.
              (see file ext/showusers/show_user_details.php).

03-Aug-2010 - Added Control file overview to DBInfo plugin.

20-Jul-2010 - Added detail view popup window to the Show Users plugin
              and also added tabs to show also all profiles and roles.

18-Jul-2010 - Added new plugin "Flashback" which shows details about the
              new 10g+ feature Flashback and recycle bin informations.

            - Init Parameter display now shows the Description field from
              the V$PARAMETER view as Tooltip

            - Added BLOCK_SIZE to tablespace overview (only for Oracle >= 9i)

            - Updated OCI8 class to 1.00, the new PHP5-only variant.

            - Rewritten all scripts to utilise my own class "sgl_functions"
              which contains several helper functions.

20-Apr-2010 - Added possibility to login as SYSOPER / SYSDBA.

            - Updated OCI8 class, jQuery (1.4.2) and jQueryUI (1.8.4)

12-Sep-2009 - Added new plugin "Top 20 Queries" which lists the most used
              queries together with a list of most invalidations per SQL.


index.php
<?php
/**
 * Login page for OIS2.
 * @package OIS2
 * @author Sascha 'SieGeL' Pfalz <php@saschapfalz.de>
 * @version 2.03 (31-Jan-2015)
 * @license http://opensource.org/licenses/gpl-license.php GNU Public License
 */
require_once('inc/nocache.inc.php');
$sysd = checkSystemData();
$lerr = (isset($_GET['LERR'])) ? intval($_GET['LERR']) : 0;
$un   = '';
if($lerr)
  {
  $errstr = '';
  switch($lerr)
    {
    case  1:
              $errstr = 'Empty username given?';
              break;
    case  2:
              $dummy  = Unserialize(Base64_Decode(strip_tags($_GET['D'])));
              $errstr = $dummy['msg'];
              $un     = $dummy['uname'];
              break;
    default:
              $errstr = 'Unknown Error '.$lerr.' - please inform the author!';
    }
  }

if(isset($_COOKIE['OIS2_LOGIN']) && $_COOKIE['OIS2_LOGIN'] != '')
  {
  $dummy = explode('|',$_COOKIE['OIS2_LOGIN']);
  $uname = $dummy[0];
  $tname = $dummy[1];
  }
else
  {
  $uname = '';
  $tname = '';
  }
if($un == '')
  {
  $un = $uname;
  }
?>
<!DOCTYPE html>
<html lang="en">
<head>
<title><?php echo(SITE_TITLE);?>: Login page</title>
<?php
require_once('inc/metatags.inc.php');
?>
<script>
$(document).ready(function() {
  $("input:submit").button();
});
</script>
</head>
<body>
<?php
if(function_exists('json_encode')==FALSE)
  {
  Error('ERROR: YOUR PHP HAS NO SUPPORT FOR "JSON" - PLEASE RECOMPILE/REINSTALL!');
  $db->disconnect();
  exit;
  }
?>
<div id="login_mask">
<h1>Welcome to <?php echo(SITE_TITLE);?></h1>
<small>Written 2009-2021 by <a href="https://www.saschapfalz.de/downloads_readme.php?FID=59&CAT=0" target="_blank" title="Click to visit my homepage to get new updates">Sascha 'SieGeL' Pfalz</a></small><br>
<br>
<form method="post" action="login_check.php" id="login_form">
<table class="tborder">
<caption>Please enter your login data:</caption>
<tr>
  <td><label for="username">Username:</label></td>
  <td><input type="text" name="USERNAME" id="username" value="<?php echo($un);?>" size="32" maxlength="32" required="required" autofocus="autofocus"></td>
</tr>
<tr>
  <td><label for="password">Password:</label></td>
  <td><input type="password" name="PASSWORD" id="password" value="" size="32" maxlength="100" required="required"></td>
</tr>
<tr>
  <td><label for="database">Database:</label></td>
  <td><select name="DATABASE" id="database" size="1">
<?php
for($i = 0; $i < count($OIS_DATABASES); $i++)
  {
  if($OIS_DATABASES[$i] == $tname)
    {
    $mysel = ' SELECTED';
    }
  else
    {
    $mysel = '';
    }
  echo("  <option value=\"".$OIS_DATABASES[$i]."\"".$mysel.">".$OIS_DATABASES[$i]."</option>\n");
  }
?>
  </select>
  </td>
</tr>
<?php
if($sysd['PRIV_CONNECT'])
  {
  echo("<tr>\n");
  echo("  <td><label for=\"conntype\">Connect as:</label></td>\n");
  echo("  <td><select name=\"CONNTYPE\" id=\"conntype\" size=\"1\"><option value=\"".OCI_DEFAULT."\">Normal</option>\n<option value=\"".OCI_SYSOPER."\">SYSOPER</option><option value=\"".OCI_SYSDBA."\">SYSDBA</option></select></td>\n");
  echo("</tr>\n");
  }
?>
</table>
<br>
<input type="submit" value="Login">
</form>
<?php
if($lerr)
  {
  echo("<div id=\"login_error\">\n");
  Error($errstr);
  echo("</div>\n");
  }
?>
</div>
<?php
require_once('inc/footer.inc.php');
?>
</body>
</html>
